crates/stages/stages/src/stages/s3/mod.rs

```diff
@@ -0,0 +1,187 @@
+mod downloader;
+pub use downloader::{fetch, Metadata};
+
+mod filelist;
+use filelist::DOWNLOAD_FILE_LIST;
+
+use reth_db::{cursor::DbCursorRW, tables, transaction::DbTxMut};
+use reth_primitives::StaticFileSegment;
+use reth_provider::{
+    BlockBodyIndicesProvider, DBProvider, StageCheckpointReader, StageCheckpointWriter,
+    StaticFileProviderFactory,
+};
+use reth_stages_api::{
+    ExecInput, ExecOutput, Stage, StageCheckpoint, StageError, StageId, UnwindInput, UnwindOutput,
+};
+
+/// S3 `StageId`
+const S3_STAGE_ID: StageId = StageId::Other("S3");
+
+/// The s3 stage.
+#[derive(Default, Debug, Clone)]
+#[non_exhaustive]
+pub struct S3Stage;
+
+impl<Provider> Stage<Provider> for S3Stage
+where
+    Provider: DBProvider<Tx: DbTxMut>
+        + StaticFileProviderFactory
+        + StageCheckpointReader
+        + StageCheckpointWriter,
+{
+    fn id(&self) -> StageId {
+        S3_STAGE_ID
+    }
+
+    fn execute(&mut self, provider: &Provider, input: ExecInput) -> Result<ExecOutput, StageError>
+    where
+        Provider: DBProvider<Tx: DbTxMut>
+            + StaticFileProviderFactory
+            + StageCheckpointReader
+            + StageCheckpointWriter,
+    {
+        let cfg = "localhost:8000";
+        let checkpoint = provider.get_stage_checkpoint(S3_STAGE_ID)?.unwrap_or_default();
+        let static_file_provider = provider.static_file_provider();
+        let mut tx_block_cursor = provider.tx_ref().cursor_write::<tables::TransactionBlocks>()?;
+
+        for block_range_files in &DOWNLOAD_FILE_LIST {
+            let (_, block_range) =
+                StaticFileSegment::parse_filename(block_range_files[0].0).expect("qed");
+
+            if block_range.end() <= checkpoint.block_number {
+                continue
+            }
+
+            for (filename, file_hash) in block_range_files {
+                if static_file_provider.directory().join(filename).exists() {
+                    // TODO: check hash if the file already exists?
+                    continue
+                }
+
+                fetch(
```

Comment:
If I read this correctly then this downloads the entire file, once we have it locally we derive all db entries from it?

for this logic we have the poll functions that we could use, this is similar to how the headers+bodies stages work